File: USTGlobal.PIP.Web/src/app/shared-module/services/capital-charge.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { Constants } from '@shared/infrastructure';
import { SharedDataService } from '@global';
import { ICapitalCharge } from '@shared/domain/ICapitalCharge';
import { ICapitalChargeMain } from '@shared/domain/ICapitalChargeMain';
import { IProjectPeriodClientPrice } from '@shared/domain';
import { IHeaderInfo } from '@shared/domain/IHeaderInfo';

@Injectable({
    providedIn: 'root'
})

export class CapitalChargeService {
    constructor(private httpClient: HttpClient, private sharedDataService: SharedDataService) {
    }

    getCapitalChargeData(pipSheetId: any): Observable<ICapitalChargeMain> {
        return this.httpClient.get<ICapitalChargeMain>(Constants.webApis.getCapitalCharge.replace('{pipSheetId}', pipSheetId));
    }

    calculations(formData: any, capitalChargeData: ICapitalChargeMain): any {
        let totalCapitalCharge = 0, periodCapitalCharge = 0, netEstimatedRevenue = 0, totalPaymentLag = 0;
        if (capitalChargeData.projectPeriodTotalDTO.length > 0) {
            // calculate month wise capital charge and total capital charge
            formData.paymentLag > 30 ? totalPaymentLag = (+formData.paymentLag - 30) *
                (capitalChargeData.capitalChargeDTO.capitalChargeDailyRate / 100) : totalPaymentLag = 0;
            capitalChargeData.projectPeriodTotalDTO.forEach(period => {
                netEstimatedRevenue = period.clientPrice - period.feesAtRisk;
                if (capitalChargeData.capitalChargeDTO.isTargetMarginPrice) {
                    periodCapitalCharge = totalPaymentLag * period.clientPrice;
                } else {
                    periodCapitalCharge = totalPaymentLag * netEstimatedRevenue;
                }
                period.capitalCharge = periodCapitalCharge;
                period.netEstimatedRevenue = netEstimatedRevenue;
                totalCapitalCharge += periodCapitalCharge;
            });
        } else {
            capitalChargeData.projectPeriodTotalDTO.forEach(period => {
                period.capitalCharge = 0;
                period.netEstimatedRevenue = 0;
            });
        }
        return { totalCapitalCharge, calculatedPeriodData: capitalChargeData.projectPeriodTotalDTO };
    }

    saveCapitalCharge(data: any): Observable<any> {
        return this.httpClient.post(Constants.webApis.saveCapitalCharge, data);
    }

    getHeader1Data(projectId: any, pipSheetId: any): Observable<IHeaderInfo> {
        return this.httpClient.get<IHeaderInfo>(Constants.webApis.getHeader1Data.replace('{projectId}', projectId)
            .replace('{pipSheetId}', pipSheetId));
    }
}

File: USTGlobal.PIP.ApplicationCore/Services/CapitalChargeService.cs
using System;
using System.Threading.Tasks;
using USTGlobal.PIP.ApplicationCore.DTOs;
using USTGlobal.PIP.ApplicationCore.Interfaces;

namespace USTGlobal.PIP.ApplicationCore.Services
{
    public class CapitalChargeService : ICapitalChargeService
    {
        private readonly ICapitalChargeRepository capitalChargeRepository;
        private readonly ISharedRepository sharedRepository;

        public CapitalChargeService(ICapitalChargeRepository capitalChargeRepository, ISharedRepository sharedRepository)
        {
            this.capitalChargeRepository = capitalChargeRepository;
            this.sharedRepository = sharedRepository;
        }

        public async Task<CapitalChargeResultSetDTO> GetCapitalCharge(int pipSheetId)
        {
            return await this.capitalChargeRepository.GetCapitalCharge(pipSheetId);
        }

        public async Task<CapitalChargeResultSetDTO> CalculateCapitalCharges(int pipSheetId)
        {
            CapitalChargeResultSetDTO capitalChargeResultSetDTO = await this.GetCapitalCharge(pipSheetId);
            if (capitalChargeResultSetDTO.capitalChargeDTO != null)
            {

                decimal capitalChargeDailyRate = Convert.ToDecimal(capitalChargeResultSetDTO.capitalChargeDTO.CapitalChargeDailyRate);
                decimal totalPaymentLag = 0, netEstimatedRevenue = 0, periodCapitalCharge = 0, totalCapitalCharge = 0;

                capitalChargeResultSetDTO.capitalChargeDTO.PipSheetId = pipSheetId;

                if (capitalChargeResultSetDTO.projectPeriodTotalDTO.Count > 0)
                {
                    if (capitalChargeResultSetDTO.capitalChargeDTO.PaymentLag > 30)
                    {
                        totalPaymentLag = (decimal)((capitalChargeResultSetDTO.capitalChargeDTO.PaymentLag - 30) * (capitalChargeDailyRate / 100));
                    }
                    else
                        totalPaymentLag = 0;

                    capitalChargeResultSetDTO.projectPeriodTotalDTO.ForEach(period =>
                    {
                        netEstimatedRevenue = period.ClientPrice - period.FeesAtRisk;
                        if (capitalChargeResultSetDTO.capitalChargeDTO.IsTargetMarginPrice)
                        {
                            periodCapitalCharge = totalPaymentLag * period.ClientPrice;
                        }
                        else
                        {
                            periodCapitalCharge = totalPaymentLag * netEstimatedRevenue;
                        }

                        period.CapitalCharge = periodCapitalCharge;
                        period.NetEstimatedRevenue = netEstimatedRevenue;
                        totalCapitalCharge += periodCapitalCharge;
                    });
                }
                else
                {
                    capitalChargeResultSetDTO.projectPeriodTotalDTO.ForEach(period =>
                    {
                        period.CapitalCharge = 0;
                        period.NetEstimatedRevenue = 0;
                    });
                }
                capitalChargeResultSetDTO.capitalChargeDTO.CapitalCharge = totalCapitalCharge;
            }
            return capitalChargeResultSetDTO;
        }

        public async Task SaveCapitalCharge(string userName, CapitalChargeResultSetDTO capitalCharge)
        {
            decimal? totalProjectCost = capitalCharge.capitalChargeDTO.TotalCostBeforeCap + capitalCharge.capitalChargeDTO.CapitalCharge;
            await this.capitalChargeRepository.SaveCapitalCharge(userName, capitalCharge, totalProjectCost);
        }
    }
}
